src/assertions.ts

```diff
@@ -104,77 +105,89 @@ export async function runAssertions({
   logProbs?: number[];
   cost?: number;
 }): Promise<GradingResult> {
-  const tokensUsed = {
-    total: 0,
-    prompt: 0,
-    completion: 0,
-  };
   if (!test.assert || test.assert.length < 1) {
-    return { pass: true, score: 1, reason: 'No assertions', tokensUsed, assertion: null };
+    return AssertionsResult.noAssertsResult();
   }
-  let totalScore = 0;
-  let totalWeight = 0;
-  let allPass = true;
-  let failedReason = '';
-  const componentResults: GradingResult[] = [];
-  const namedScores: Record<string, number> = {};
-
-  await async.forEachOfLimit(test.assert, ASSERTIONS_MAX_CONCURRENCY, async (assertion, index) => {
-    if (assertion.type.startsWith('select-')) {
-      // Select-type assertions are handled separately because they depend on multiple outputs.
-      return;
-    }
-    const weight = assertion.weight ?? 1;
-    totalWeight += weight;
-    const result = await runAssertion({
-      prompt,
-      provider,
-      assertion,
-      test,
-      output,
-      latencyMs,
-      logProbs,
-      cost,
-    });
-    totalScore += result.score * weight;
-    componentResults[Number(index)] = result;
-    if (assertion.metric) {
-      namedScores[assertion.metric] = (namedScores[assertion.metric] || 0) + result.score;
-    }
-    if (result.tokensUsed) {
-      tokensUsed.total += result.tokensUsed.total;
-      tokensUsed.prompt += result.tokensUsed.prompt;
-      tokensUsed.completion += result.tokensUsed.completion;
-    }
-    if (!result.pass) {
-      allPass = false;
-      failedReason = result.reason;
-      if (process.env.PROMPTFOO_SHORT_CIRCUIT_TEST_FAILURES) {
-        throw new Error(result.reason);
+
+  const mainAssertResult = new AssertionsResult({
+    threshold: test.threshold,
+  });
+  const subAssertResults: AssertionsResult[] = [];
+  const asserts: {
+    assertion: Assertion;
+    assertResult: AssertionsResult;
+    index: number;
+  }[] = test.assert
```

Comment:
This is all flattened so we can support maximum concurrency when the assertions are actually implemented.

By making this array flat we are more likely to fill `ASSERTIONS_MAX_CONCURRENCY` vs partially filling it.